src/pycram/probabilistic_costmap.py

```diff
@@ -0,0 +1,166 @@
+    """
+    The variable for the y-axis (width) in meters.
+    """
+
+    costmap: Costmap
+    """
+    The legacy costmap.
+    """
+
+    origin: Pose
+    """
+    The origin of the costmap.
+    """
+
+    size: Quantity
+    """
+    The side length of the costmap. The costmap is a square.
+    """
+
+    distribution: Optional[ProbabilisticCircuit] = None
+    """
+    The distribution associated with the costmap.
+    """
+
+    def __init__(self, origin: Pose,
+                 size: Quantity = 2 * meter,
+                 max_cells = 10000,
+                 costmap_type: Type[Costmap] = OccupancyCostmap,
+                 world: Optional[World] = None):
+
+        self.world = world if world else World.current_world
+        self.origin = origin
+        self.size = size
+
+        # calculate the number of cells per axis
+        number_of_cells = int(np.sqrt(max_cells))
+        resolution = self.size.to(meter) / number_of_cells
+
+        if costmap_type == OccupancyCostmap:
+            self.costmap = OccupancyCostmap(
+                origin=self.origin,
+                distance_to_obstacle=0.2,
+                size=number_of_cells,
+                resolution=resolution.magnitude,
+                from_ros=False,
+                world = self.world)
+        elif costmap_type == VisibilityCostmap:
+            self.costmap = VisibilityCostmap(
+                min_height=0, max_height=2, size=number_of_cells, resolution=resolution.magnitude,
+                origin=self.origin, world=self.world)
+        else:
+            raise NotImplementedError(f"Unknown costmap type {costmap_type}")
+
+        self.create_distribution()
+
+    @cached_property
+    def publisher(self):
+        return create_publisher("/pycram/viz_marker", MarkerArray, queue_size=10)
+
+    def create_event_from_map(self) -> Event:
+        """
+        :return: The event that is encoded by the costmaps map.
+        """
+        area = Event()
+        for rectangle in self.costmap.partitioning_rectangles():
+            rectangle.translate(self.origin.position.x, self.origin.position.y)
+            area.simple_sets.add(SimpleEvent({self.x: closed_open(rectangle.x_lower, rectangle.x_upper),
+                                                    self.y: closed_open(rectangle.y_lower, rectangle.y_upper)}))
+        return area
+
+    def create_distribution(self):
+        """
+        Create a probabilistic circuit from the costmap.
+        """
+        # self.distribution = fully_factorized([self.x, self.y], {self.x: self.origin.position.x,
+        #                                                     self.y: self.origin.position.y},
+        #                                      {self.x: 1, self.y: 1})
+        # self.distribution, _ = self.distribution.conditional(self.create_event_from_map())
+        self.distribution = uniform_measure_of_event(self.create_event_from_map())
+
+    def sample_to_pose(self, sample: np.ndarray) -> Pose:
```

Comment:
There is a PoseGenerator class for this which also allows different orientation generators. That would be better to use